timemachine/cpp/src/bd_exchange_move.cu

```diff
@@ -217,21 +218,39 @@ void BDExchangeMove<RealType>::move(
             d_lse_exp_sum_after_.data,
             stream);
 
-        k_attempt_exchange_move<RealType><<<1, 1, 0, stream>>>(
-            N,
-            d_mh_noise_.data + (step * batch_size_),
+        k_select_first_valid_move<RealType><<<1, min(512, batch_size_), 0, stream>>>(
+            num_proposals_per_move_,
+            num_target_mols_,
+            batch_size_,
+            d_noise_offset_.data,
+            d_samples_.data,
             d_lse_max_before_.data,
             d_lse_exp_sum_before_.data,
             d_lse_max_after_.data,
             d_lse_exp_sum_after_.data,
-            d_target_mol_offsets_.data,
+            d_mh_noise_.data,
+            d_selected_sample_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        k_accepted_exchange_move<<<1, 1, 0, stream>>>(
+            batch_size_,
+            mol_size_,
+            d_selected_sample_.data,
             d_samples_.data,
+            d_target_mol_offsets_.data,
             d_intermediate_coords_.data,
             d_coords,
-            d_num_accepted_.data);
+            d_num_accepted_.data,
+            d_noise_offset_.data);
         gpuErrchk(cudaPeekAtLastError());
-        num_attempted_++;
+        gpuErrchk(cudaMemcpyAsync(
+            p_noise_offset_.data, d_noise_offset_.data, d_noise_offset_.size(), cudaMemcpyDeviceToHost, stream));
+        step++;
+        // Synchronize to get the new offset
+        gpuErrchk(cudaStreamSynchronize(stream));
     }
+    // Number of attempts is always the batch size
```

Comment:
nit: is always incremented by the `batch_size`

Reply:
Fixed in https://github.com/proteneer/timemachine/pull/1274/commits/3821557f46e091272e0297d24437e906375740fe